### ISSUE TITLE
Improving usage/detection of URL prefix in deployed environments

### ISSUE THREAD

**Issue by banesullivan** [jupyterhub/jupyter-server-proxy#402]
## TL;DR

Is there a better way to use `jupyter-server-proxy` / know the proxy URL in deployed environments such as MyBinder, JupyterHub, and SageMaker?

Users on a handful of projects where I leverage `jupyter-server-proxy` struggle to get the correct URL prefix in managed deployed environments; often it's just `'/{JUPYTERHUB_SERVICE_PREFIX}/proxy/<port>` but other times it can be more nuanced. 

Are there any options or existing efforts to alleviate this pain point for users?

## Use case

I've been leveraging `jupyter-server-proxy` on a handful of projects in a similar fashion, which, may be different from the use cases `jupyter-server-proxy` is designed for. I'd appreciate clarification on if this usage scenario is well-supported and if there are ways to improve my usage of `jupyter-server-proxy` for these use cases to relieve configuration pain for our users.

In each of these projects, we spin up one or many web servers within the Python kernel (at runtime) to then access from the client-side Jupyter notebook. Two examples:

1. **Serving tiles for large images to visualize interactively on a map.** The user launches a tile server (with their configurations defined at runtime) for a given "image" instance in their kernel. We then combine Jupyter widgets like `ipyleaflet` with the REST tile endpoint we've launched on an arbitrary port on the host machine. We need the client browser to access that server to fetch the tiles. See https://github.com/girder/large_image/pull/1065 and https://github.com/banesullivan/localtileserver

<img width="1306" alt="Screen Shot 2023-04-24 at 11 14 57 AM" src="https://user-images.githubusercontent.com/22067021/234069149-9d479dea-707e-4d5f-af2c-ceaa29c1a173.png">

3. **Custom Jupyter "widgets" as an iFrame.** In [PyVista](https://github.com/pyvista/pyvista), we serve a single web application that links to individual 3D plots the user creates in their kernel. We need full access to the port this application runs on and display the application as an iFrame in the Jupyter Notebook so to the user it's just like any other widget. This web app must be launched within the same thread as the user's Python kernel's main thread. More details [here](https://docs.pyvista.org/version/stable/user-guide/jupyter/trame.html#trame-jupyter) and [here](https://discourse.vtk.org/t/pyvista-trame-jupyter-3d-visualization/10610)

![Screen Shot 2023-01-31 at 12 02 34 PM](https://user-images.githubusercontent.com/22067021/234069269-d57f1641-f68c-4174-a74b-e951fa8ed159.png)

The primary constraint shared in these use cases is that the web server must be a part of the user's Python kernel process: we cannot launch the server ahead of time, and we cannot share servers across Python sessions (notebooks).

Using `jupyter-server-proxy` has gone excellently for these use cases with only minor issues around end-user configuration and request timeouts -- which is what I hope to optimize and solicit as improved support for in `jupyter-server-proxy`.

## Help!

1. Are these use cases well-supported by `jupyter-server-proxy`? I'm asking because the documentation focuses on launching a single standalone web application on the Jupyter Server rather than this launching of arbitrary servers from within a user's Python session.

2. Are there any ways to improve knowing the URL prefix ahead of time in deployed environments (like MyBinder or SageMaker) to prevent users from having to debug different prefixes in `/<prefix>/proxy/<port>` for these use cases?

### Common prefix scenarios

- MyBinder: `f"{os.environ['JUPYTERHUB_SERVICE_PREFIX']}/proxy/{{port}}"`
- SageMaker: `f"studiolab/default/jupyter/proxy/{{port}}"`
- Colab: actually, `jupyter-server-proxy` isn't needed as `localhost` is remapped by Colab ([reference](https://colab.research.google.com/notebooks/snippets/advanced_outputs.ipynb))

## Related

Pinging a few folks who may want to follow this issue: - @giswqs, @akaszynski

- https://github.com/pyvista/pyvista/issues/3987
- https://github.com/banesullivan/localtileserver/issues/134
- https://github.com/banesullivan/localtileserver/issues/128
- https://github.com/banesullivan/localtileserver/issues/95
- https://github.com/banesullivan/localtileserver/issues/66

**Comment by giswqs** [comment]
It would be nice to have a solution that works across platforms, such as MyBinder, Google Colab, JupyterHub, Amazon SageMaker, Microsoft Planetary Computer.

**Comment by manics** [comment]
`JUPYTERHUB_SERVICE_PREFIX` is the standard way to get the base prefix with JupyterHub. As far as I know SageMaker and Colab aren't open-source, so I don't think there's much we can do. However if there's a "well known" standard way to get the prefix we could consider supporting it. Are you aware of any documention?

If not, it might be better to move this discussion to the Jupyter Community Forum https://discourse.jupyter.org/ where people from across the Jupyter ecosystem hang out, since they may have run into this issue with other extensions.

**Comment by ryanlovett** [comment]
> Are these use cases well-supported by jupyter-server-proxy? I'm asking because the documentation focuses on launching a single standalone web application on the Jupyter Server rather than this launching of arbitrary servers from within a user's Python session.

Perhaps it is not well-supported right now, because that hasn't been the focus, but that doesn't mean it couldn't be. I'll add some initial thoughts below, but as @manics said it might be worthwhile to discuss on discourse.

> Are there any ways to improve knowing the URL prefix ahead of time in deployed environments (like MyBinder or SageMaker) to prevent users from having to debug different prefixes in /<prefix>/proxy/<port> for these use cases?

The only way to know the path_info of the managed process (the part after the path to the user's server) in advance is with a named server proxy like those for [RStudio](https://github.com/jupyterhub/jupyter-rsession-proxy), [VSCode](https://github.com/betatim/vscode-binder/), [noVNC](https://github.com/jupyterhub/jupyter-remote-desktop-proxy), etc., although these are for single instances. These are made available via entry points where the configuration is known in advance. Right now all existing entry points are loaded when the extension is loaded.

I imagine that there could be a jupyter-server-proxy HTTP API to either create new named services, or to reload any new entry points that have been created since the extension was loaded. This would enable the user to run some code to start a proxied service with an endpoint of their choosing.

**Comment by banesullivan** [comment]
@manics and @ryanlovett, thank you both for this initial feedback! I'll try to re-pose some of this discussion on Discourse.

> I imagine that there could be a jupyter-server-proxy HTTP API to either create new named services, or to reload any new entry points that have been created since the extension was loaded. This would enable the user to run some code to start a proxied service with an endpoint of their choosing.

This is a neat idea. I suppose I confused on why something must be a new named service to know the `path_info`... is this not available for the main `/proxy/<port>` service?
